packages/api-contract/src/Abi/index.ts

```diff
@@ -82,27 +108,38 @@ export class Abi {
   readonly messages: AbiMessage[];
   readonly metadata: ContractMetadataLatest;
   readonly registry: Registry;
+  readonly environment?: { [x: string]: unknown }[];
 
   constructor (abiJson: Record<string, unknown> | string, chainProperties?: ChainProperties) {
-    [this.json, this.registry, this.metadata, this.info] = parseJson(
+    [this.json, this.registry, this.metadata, this.info, this.environment] = parseJson(
       isString(abiJson)
         ? JSON.parse(abiJson) as Record<string, unknown>
         : abiJson,
       chainProperties
     );
-    this.constructors = this.metadata.spec.constructors.map((spec: ContractConstructorSpecLatest, index) =>
-      this.#createMessage(spec, index, {
+    this.constructors = this.metadata.spec.constructors.map((spec: ContractConstructorSpecLatest, index) => {
+      const isDefault = 'default' in spec ? spec.default.isTrue : undefined;
+      const typeSpec = 'returnType' in spec ? spec.returnType.unwrapOr(null) : null;
```

Comment:
Since we are always dealing with the latest spec here, it will be available. (We always only work with latest, hence the conversions - this means we don't need specific logic for each intermediate version as added here)

(For V3 -> V4 conversions, it would be false, the default if not supplied)

Reply:
I'm doing something wrong i think. I added those checks because of runtime errors when working with older metadata.  So no, they weren't always available.
Should I add the fields with `undefined` here? https://github.com/polkadot-js/api/blob/8280bed13b7ee2f9dc6fa11ca7b9772cfba6c695/packages/api-contract/src/Abi/toV4.ts#L10

And then try to read the values from spec?

Reply:
They don't need to be added as undefined - when you create and it is not supplied, it will fill in with the default values for the specified types. (aka it will be all-0 values).

An example with `default` - since it is not supplied, it will just be `false`.

You only need to supply them explicitly if you wish to supply specific values. (In this case just leaving them out and accepting the defaults should be fine)